proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,306 @@
+
+
+## Non-Goals
+
+* There's a possiblity of *one-to-many* where a single virtual cluster maps to several physical, presenting them as if they were one. However, supporting transactions across two or more physical clusters would be difficult.  This use-case is out-of-scope.
+* It might be possible for a virtual cluster to present a subset of a physical cluster's brokers as virtual brokers by clevery exposing only those brokers that host topic partitions belonging to that virtual cluster.  This might be advantagous in the case where the physical cluster comprises a very large number of brokers.  We won't cosnider this use-case for now. 
+
+
+## Current situation
+
+Kroxyliciousis currently limited to exposing a single broker of a single cluster.
+
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
+
+```yaml
+- name: endpoint name
+  address: bind address/port
+  type: PLAIN|TLS
+```
+
+#### Virtual Clusters
+
+Conceptually the _virtual cluster_ represents the cluster that the kafka client connects to.  The virtual clusters list specifies
+all the kafka clusters are being presented by Kroxylicious. 
+
+A _virtual clusster_ references exactly one _physical cluster_. 
+
+A _virtual cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the physical cluster.
+
+A _virtual cluster_ enumerates the virtual brokers that comprise the virtual cluster.
+
+##### Virtual Brokers
+
+Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
+for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.
+
+Each virtual broker references exactly one physical broker which must exist within the referenced physical cluster.  It is an error if there is not a
+one-to-one correspondence between downstream and upstream brokers.
+
+TLS key material may be provided either at the virtual cluster level or at the individual broker level.
+
+
+```yaml
+- name: my-public-cluster
```

Comment:
I found it confusing when the yaml didn't include the `virtualClusters` key. I think it's more clear if it's something like this:
```yaml
...
virtualClusters:
- name: my-public-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-downstream--filter-chain
  tls:
    key:
    cert:
  brokers:
 name: my-public-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com
    physicalBrokerRef: broker-1
    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
    ...
```